### PR TITLE
Blog - azure service fabric setup

### PR DESCRIPTION
**Pull request by colin-higgins**
Guide on how to enable the .NET Tracer in Azure Service Fabric.

[Preview link](https://github.com/DataDog/dd-trace-dotnet/blob/blogs/azure-service-fabric-setup/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md)

### CODE REVIEW

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+This article contains guidance for enabling APM on applications built with [Microsoft Azure Service Fabric](https://azure.microsoft.com/services/service-fabric/). The quickstart project contains a single application with multiple services demonstrating the basic concepts of service communication and use of reliable dictionaries.
+
+For a more in depth tour with the sample used for this demo, visit: [Service Fabric .NET quickstart](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-quickstart-dotnet)
+
+More info on Service Fabric:
+ - [Documentation](https://docs.microsoft.com/azure/service-fabric/)
+ - [Service Fabric sample projects](https://azure.microsoft.com/resources/samples/?service=service-fabric)
+ - [Service Fabric open source home repo](https://github.com/azure/service-fabric)
+ 
+There are two particular requirements for enabling Datadog in Azure Service Fabric: Deploying the datadog agent, and installing the .NET Tracer
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:17]:
```suggestion
There are two particular requirements for enabling Datadog in Azure Service Fabric: Deploying the Datadog Agent, and installing the .NET Tracer.
```

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
+
+# Datadog .NET APM in Service Fabric
+
+*This guidance is in an alpha state, and is not officially supported*
+
+*However, if you try it, we would love to hear your feedback.*
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:6]: We should add a feedback link.

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+This article contains guidance for enabling APM on applications built with [Microsoft Azure Service Fabric](https://azure.microsoft.com/services/service-fabric/). The quickstart project contains a single application with multiple services demonstrating the basic concepts of service communication and use of reliable dictionaries.
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:8]:
```suggestion
This article contains guidance for enabling APM on .NET applications built with [Microsoft Azure Service Fabric](https://azure.microsoft.com/services/service-fabric/). The quickstart project contains a single application with multiple services demonstrating the basic concepts of service communication and use of reliable dictionaries.
```

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+For a more in depth tour with the sample used for this demo, visit: [Service Fabric .NET quickstart](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-quickstart-dotnet)
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:10]:
```suggestion
For a more in-depth tour with the sample used for this demo, visit: [Service Fabric .NET quickstart](https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-quickstart-dotnet).
```

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+More info on Service Fabric:
+ - [Documentation](https://docs.microsoft.com/azure/service-fabric/)
+ - [Service Fabric sample projects](https://azure.microsoft.com/resources/samples/?service=service-fabric)
+ - [Service Fabric open source home repo](https://github.com/azure/service-fabric)
+ 
+There are two particular requirements for enabling Datadog in Azure Service Fabric: Deploying the datadog agent, and installing the .NET Tracer
+
+The datadog agent is a standalone docker container which exists as a service.
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:19]:
```suggestion
The Datadog Agent is a standalone Docker container which exists as a service.
```

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+The .NET tracer requires including some scripts with a deployed service to install the .NET Tracer on your clusters.
+ 
+## Create a service fabric application
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:23]:
```suggestion
## Create a Service Fabric application
```

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+If you already have a service fabric application, you can skip ahead to `Setting up the datadog agent`.
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:25]:
```suggestion
If you already have a Service Fabric application, you can skip ahead to [_Setting up the Datadog Agent_](#setting-up-the-datadog-agent).
```

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+Otherwise, here is one way to setup a Service Fabric Application in Visual Studio:
+
+![Create a service fabric project](https://user-images.githubusercontent.com/1801443/93098850-5079fd80-f675-11ea-90d6-7573b7faef68.png)
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:29]:
```suggestion
![Create a Service Fabric project](https://user-images.githubusercontent.com/1801443/93098850-5079fd80-f675-11ea-90d6-7573b7faef68.png)
```

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+
+Start with a stateless ASP.NET Core application, and use MVC as a template.
+
+![Create a stateless ASP.NET Core application](https://user-images.githubusercontent.com/1801443/93099063-959e2f80-f675-11ea-805c-eb627e2b9e53.png)
+
+## Setting up the Datadog agent
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:35]:
```suggestion
## Setting up the Datadog Agent
```

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+
+Add a new Service Fabric Service to the solution.
+
+![Add new Service Fabric Service](https://user-images.githubusercontent.com/1801443/93102030-04c95300-f679-11ea-89f2-1de6160b5bc2.png)
+
+![Setup the agent container](https://user-images.githubusercontent.com/1801443/93107331-73111400-f67f-11ea-9a5e-06094e775177.png)
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:41]:
```suggestion
![Setup the Datadog Agent container](https://user-images.githubusercontent.com/1801443/93107331-73111400-f67f-11ea-9a5e-06094e775177.png)
```

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+    <Endpoints>
+      <Endpoint Name="DatadogTypeEndpoint" UriScheme="http" Port="5002" Protocol="http"/>
+      <Endpoint Name="DatadogTraceEndpoint" UriScheme="http" Port="8126" Protocol="http"/>
+      <Endpoint Name="DatadogStatsEndpoint" UriScheme="udp" Port="8125" Protocol="udp"/>
+    </Endpoints>
+  </Resources>
+</ServiceManifest>
+```
+
+Add the corresponding port bindings to the ServiceManfestImport within ApplicationManifest.xml
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:75]:
```suggestion
Add the corresponding port bindings to the `ServiceManfestImport` in `ApplicationManifest.xml`.
```

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+
+## Setting up the Datadog agent
+
+Add a new Service Fabric Service to the solution.
+
+![Add new Service Fabric Service](https://user-images.githubusercontent.com/1801443/93102030-04c95300-f679-11ea-89f2-1de6160b5bc2.png)
+
+![Setup the agent container](https://user-images.githubusercontent.com/1801443/93107331-73111400-f67f-11ea-9a5e-06094e775177.png)
+
+Replace the ServiceManifest.xml with this, being sure to replace `API_KEY_GOES_HERE` with your Datadog API key:
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:43]:
```suggestion
Replace the `ServiceManifest.xml` with the following text. Make sure you replace `API_KEY_GOES_HERE` with your Datadog API key:
```

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
++       <PortBinding ContainerPort="8125" EndpointRef="DatadogStatsEndpoint" />
++     </ContainerHostPolicies>
++   </Policies>
+  </ServiceManifestImport>
+```
+
+## Installing the .NET Tracer
+
+The tracer requires machine administrator permissions.
+Add the SetupAdminUser to the Principals section in the ApplicationManifest.xml. If the Principals section is missing, add it.
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:94]:
```suggestion
Add the `SetupAdminUser` to the `Principals` element in the `ApplicationManifest.xml` file. If the `Principals` section is missing, add it.
```

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+      <User Name="SetupAdminUser">
+        <MemberOf>
+          <SystemGroup Name="Administrators" />
+        </MemberOf>
+      </User>
+    </Users>
+  </Principals>
+```
+
+Within the ServiceManifestImport of the service responsible for deploying the tracer to the cluster, give the Setup script SetupAdminUser as the executing user.
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:108]:
```suggestion
In the `ServiceManifestImport` element of the service responsible for deploying the tracer to the cluster, set `SetupAdminUser` as the executing user for the `Setup` script.
```

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+  <ServiceManifestImport>
+    <ServiceManifestRef ServiceManifestName="ServiceThatDeploysDatadogTracerPkg" ServiceManifestVersion="1.0.0" />
+    <ConfigOverrides />
++   <Policies>
++     <RunAsPolicy CodePackageRef="Code" UserRef="SetupAdminUser" EntryPointType="Setup" />
++   </Policies>
+  </ServiceManifestImport>
+```  
+
+Within the ServiceManifest.xml of the service responsible for deploying the Datadog Tracer, add the reference to the install script.
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:120]:
```suggestion
In the `ServiceManifest.xml` file of the service responsible for deploying the Datadog Tracer, add the reference to the install script.
```

Comment (lucaspimentel) [reply docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:120]: > Datadog Tracer

Did you mean Datadog Agent or .NET Tracer?

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 16]
```diff
@@ -0,0 +1,156 @@
...
+    <EntryPoint>
+      <ExeHost>
+        <Program>ServiceThatDeploysDatadogTracer.exe</Program>
+        <WorkingFolder>CodePackage</WorkingFolder>
+      </ExeHost>
+    </EntryPoint>
+  </CodePackage>
+```
+
+Include the DatadogInstall.bat and DatadogInstall.ps1 scripts in the project responsible for deploying the tracer.
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:138]:
```suggestion
Include the `DatadogInstall.bat` and `DatadogInstall.ps1` scripts in the project responsible for deploying the tracer.
```

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 17]
```diff
@@ -0,0 +1,156 @@
...
+Ensure that in the properties of both scripts, you set that they are copied to the output directory.
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:139]:
```suggestion
Configure both files to be copied to the output directory.
```

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+
+The latest representation of this install process is here: https://github.com/DataDog/azureservicefabric-dotnet-tracing-sample
+ - [Batch Script](https://github.com/DataDog/azureservicefabric-dotnet-tracing-sample/blob/master/VotingWeb/DatadogInstall.bat)
+ - [Powershell Script](https://github.com/DataDog/azureservicefabric-dotnet-tracing-sample/blob/master/VotingWeb/DatadogInstall.ps1)
+
+
+### That's all folks
+---
+
+The next time you deploy this application to the cluster, your application should start sending traces through to Datadog.
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:151]:
```suggestion
The next time you deploy this application to the cluster, your application should start sending traces to Datadog.
```

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 19]
```diff
@@ -0,0 +1,156 @@
...
+This also enables custom statistics through the DogStatsD client and custom traces through the Datadog.Trace library.
```
Comment (lucaspimentel) [inline docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:152]:
```suggestion
This also enables custom metrics through DogStatsD and custom traces through the `Datadog.Trace` library.
```

Comment (lucaspimentel) [reply docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:152]: We should link to both NuGet packages.

**Review by lucaspimentel — COMMENTED** [review]
Left some suggestions, mostly for brand capitalization.

File: docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md (hunk 19 above)
Comment (macrogreg) [reply docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:152]: Could we please call the directory "blog/" rather than "blog*s*"?

Comment (macrogreg) [reply docs/blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:152]: Sorry, never mind, confused branch and folder :)

File: blog/2020-09-15-datadog-apm-in-azure-service-fabric.md [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+ - [Batch Script](https://github.com/DataDog/azureservicefabric-dotnet-tracing-sample/blob/master/VotingWeb/DatadogInstall.bat)
+ - [Powershell Script](https://github.com/DataDog/azureservicefabric-dotnet-tracing-sample/blob/master/VotingWeb/DatadogInstall.ps1)
+
+Configure both files to be copied to the output directory.
+
+![Copy to output directory](https://user-images.githubusercontent.com/1801443/93110062-d05a9480-f682-11ea-8fb4-7b266f576f68.png)
+
+
+Deploy your application to the cluster, generate some load against it, and you should start seeing your infrastructure and traces coming through to your Datadog account.
+
```
Comment (macrogreg) [inline blog/2020-09-15-datadog-apm-in-azure-service-fabric.md:181]: ...to your Datadog account. For example:

- View [distributed traces](https://app.datadoghq.com/apm/traces) in the Datadog portal.
- View a [host map](https://app.datadoghq.com/infrastructure/map) in the Datadog portal.

**Review by macrogreg — APPROVED** [review]

**Review by lucaspimentel — COMMENTED** [review]
I got bit by github's UI again. I added all this comments and never submitted the review.

**Pull request merged**